File: frontend/src/lib/machineUtility.ts
/**
 * Machine Utility Consumption Mapping — TypeScript types, label maps, API client.
 *
 * Two-layer design:
 *   MachineUtilityMapping     master config: machine ↔ meter link + rated consumption
 *   MachineConsumptionRecord  operational log: actual consumption per machine/shift
 *
 * KPIs:
 *   electricity_per_runtime_h  kWh / runtime hour
 *   water_per_batch            m³  / batch
 *   air_per_batch              Nm³ / batch
 *   standby_pct                standby share of total consumption %
 *   cost_per_batch             total_cost / batch count
 *   avg_variance_pct           deviation from nominal rate
 */
import { apiClient } from "./api";

// ── Enums & label maps ────────────────────────────────────────────────────────

export type UtilityType =
  | "ELECTRICITY"
  | "WATER"
  | "SOFT_WATER"
  | "PROCESS_WATER"
  | "STEAM"
  | "COMPRESSED_AIR"
  | "NATURAL_GAS"
  | "SOLAR"
  | "DIESEL"
  | "CHILLED_WATER"
  | "OTHER";

export const UTILITY_TYPE_LABELS: Record<UtilityType, string> = {
  ELECTRICITY:    "Electricity",
  WATER:          "Water",
  SOFT_WATER:     "Soft Water",
  PROCESS_WATER:  "Process Water",
  STEAM:          "Steam",
  COMPRESSED_AIR: "Compressed Air",
  NATURAL_GAS:    "Natural Gas",
  SOLAR:          "Solar",
  DIESEL:         "Diesel",
  CHILLED_WATER:  "Chilled Water",
  OTHER:          "Other",
};

export const UTILITY_TYPE_COLORS: Record<UtilityType, string> = {
  ELECTRICITY:    "#fbbf24",
  WATER:          "#60a5fa",
  SOFT_WATER:     "#93c5fd",
  PROCESS_WATER:  "#3b82f6",
  STEAM:          "#f97316",
  COMPRESSED_AIR: "#22d3ee",
  NATURAL_GAS:    "#a78bfa",
  SOLAR:          "#facc15",
  DIESEL:         "#94a3b8",
  CHILLED_WATER:  "#67e8f9",
  OTHER:          "#6b7280",
};

export const SOURCE_METHOD_LABELS: Record<string, string> = {
  MANUAL:     "Manual Entry",
  ESTIMATED:  "Estimated",
  CALCULATED: "Calculated",
  IOT:        "IoT / SCADA",
  API:        "External API",
  IMPORTED:   "CSV Import",
};

// ── Mapping master types ───────────────────────────────────────────────────────

export interface MachineUtilityMapping {
  id:              string;
  machine_ref:     string;
  machine_name:    string;
  work_center_id?: string | null;
  department?:     string | null;
  production_line?: string | null;
  building_area?:  string | null;
  utility_type:    UtilityType;
  device_id?:      string | null;
  device_name?:    string | null;
  asset_id?:       string | null;
  allocation_pct?: number | null;
  rated_consumption?: number | null;
  rated_unit?:     string | null;
  is_active:       boolean;
  notes?:          string | null;
  created_at?:     string | null;
  updated_at?:     string | null;
}

export type MachineUtilityMappingCreate = Omit<MachineUtilityMapping,
  "id" | "device_name" | "created_at" | "updated_at"
>;

export type MachineUtilityMappingUpdate = Partial<Omit<MachineUtilityMapping,
  "id" | "machine_ref" | "device_name" | "created_at" | "updated_at"
>>;

// ── Consumption record types ───────────────────────────────────────────────────

export interface MachineConsumptionRecord {
  id:              string;
  record_no:       string;
  mapping_id?:     string | null;
  work_center_id:  string;
  machine_ref:     string;
  machine_name?:   string | null;
  utility_type:    UtilityType;
  device_id?:      string | null;
  device_name?:    string | null;

  record_date:     string;
  shift_ref?:      string | null;
  department?:     string | null;
  production_line?: string | null;

  source_method:   string;
  is_estimated:    boolean;
  is_standby:      boolean;

  runtime_hours?:  number | null;

  nominal_rate?:   number | null;
  nominal_unit?:   string | null;

  actual_consumption?: number | null;
  actual_unit?:    string | null;
  variance_pct?:   number | null;

  production_order_id?: string | null;
  production_order_no?: string | null;
  batch_no?:       string | null;
  batch_lot_id?:   string | null;

  cost_rate?:      number | null;
  total_cost?:     number | null;
  currency_code?:  string | null;

  is_anomaly:      boolean;
  anomaly_note?:   string | null;
  notes?:          string | null;

  entered_by_id?:   string | null;
  entered_by_name?: string | null;

  created_at?: string | null;
  updated_at?: string | null;
}

export type MachineConsumptionRecordCreate = Omit<MachineConsumptionRecord,
  "id" | "record_no" | "machine_ref" | "machine_name" | "device_name"
  | "production_order_no" | "entered_by_id" | "entered_by_name"
  | "created_at" | "updated_at"
>;

export type MachineConsumptionRecordUpdate = Partial<Omit<MachineConsumptionRecord,
  "id" | "record_no" | "work_center_id" | "machine_ref" | "machine_name"
  | "device_name" | "production_order_no" | "entered_by_id" | "entered_by_name"
  | "created_at" | "updated_at"
>>;

// ── KPI & analytics types ─────────────────────────────────────────────────────

export interface MachineUtilityKPIs {
  date_from?: string | null;
  date_to?:   string | null;
  machine_ref?: string | null;
  utility_type?: string | null;

  total_consumption:          number;
  total_runtime_hours?:       number | null;
  total_standby_consumption?: number | null;
  consumption_unit?:          string | null;

  electricity_per_runtime_h?: number | null;
  water_per_batch?:           number | null;
  air_per_batch?:             number | null;
  standby_pct?:               number | null;
  cost_per_batch?:            number | null;
  cost_per_unit?:             number | null;

  total_cost?:     number | null;
  avg_cost_rate?:  number | null;
  avg_variance_pct?: number | null;

  record_count:    number;
  batch_count:     number;
  order_count:     number;
  anomaly_count:   number;

  flag_high_variance:  boolean;
  flag_standby_waste:  boolean;
}

export interface MachineUtilityTrendPoint {
  date:              string;
  machine_ref?:      string | null;
  utility_type?:     string | null;
  total_consumption: number;
  standby:           number;
  runtime_hours:     number;
  total_cost?:       number | null;
  batch_count:       number;
  record_count:      number;
}

export interface MachineUtilityBreakdownRow {
  group_key?:        string | null;
  label:             string;
  total_consumption: number;
  total_cost?:       number | null;
  pct_of_total?:     number | null;
  runtime_hours?:    number | null;
  record_count:      number;
}

export interface MachineUtilityBreakdown {
  dimension:    string;
  rows:         MachineUtilityBreakdownRow[];
  grand_total:  number;
}

// ── Dropdown types ─────────────────────────────────────────────────────────────

export interface WorkCenterDropdown {
  id:             string;
  work_center_id: string;
  name:           string;
  department?:    string | null;
  type:           string;
}

export interface MappingDropdown {
  id:           string;
  machine_ref:  string;
  machine_name: string;
  utility_type: UtilityType;
  rated_consumption?: number | null;
  rated_unit?:  string | null;
}

export interface DeviceDropdown {
  id:              string;
  device_code:     string;
  name:            string;
  utility_type:    string;
  unit_of_measure?: string | null;
}

export interface ProductionOrderDropdown {
  id:       string;
  order_no: string;
  batch_no?: string | null;
  status:   string;
}

// ── Filters ────────────────────────────────────────────────────────────────────

export interface MachineUtilityFilters {
  work_center_id?:      string;
  machine_ref?:         string;
  utility_type?:        UtilityType;
  production_line?:     string;
  department?:          string;
  shift_ref?:           string;
  date_from?:           string;
  date_to?:             string;
  batch_no?:            string;
  production_order_id?: string;
  is_standby?:          boolean;
  is_anomaly?:          boolean;
  is_estimated?:        boolean;
  skip?:                number;
  limit?:               number;
}

// ── CSV download ───────────────────────────────────────────────────────────────

export async function downloadMachineUtilityCsv(
  params: Record<string, string | number | boolean | undefined>,
  filename = "machine_consumption.csv",
): Promise<void> {
  const base  = process.env.NEXT_PUBLIC_API_URL ?? "";
  const token = typeof window !== "undefined" ? localStorage.getItem("access_token") : null;
  const qs    = new URLSearchParams(
    Object.entries(params)
      .filter(([, v]) => v !== undefined && v !== "")
      .map(([k, v]) => [k, String(v)]),
  ).toString();
  const res = await fetch(
    `${base}/api/v1/machine-utility/records/export/csv${qs ? "?" + qs : ""}`,
    { headers: token ? { Authorization: `Bearer ${token}` } : {} },
  );
  if (!res.ok) throw new Error(`Export failed (${res.status})`);
  const blob = await res.blob();
  const url  = URL.createObjectURL(blob);
  const a    = document.createElement("a");
  a.href = url; a.download = filename;
  document.body.appendChild(a); a.click();
  document.body.removeChild(a); URL.revokeObjectURL(url);
}

// ── API client ─────────────────────────────────────────────────────────────────

export const machineUtilityApi = {
  // ── Dropdowns
  async workCenters(): Promise<WorkCenterDropdown[]> {
    const r = await apiClient.get<WorkCenterDropdown[]>("/api/v1/machine-utility/work-centers");
    return r.data;
  },
  async devices(utilityType?: UtilityType): Promise<DeviceDropdown[]> {
    const r = await apiClient.get<DeviceDropdown[]>("/api/v1/machine-utility/devices",
      { params: utilityType ? { utility_type: utilityType } : {} });
    return r.data;
  },
  async productionOrders(): Promise<ProductionOrderDropdown[]> {
    const r = await apiClient.get<ProductionOrderDropdown[]>("/api/v1/machine-utility/production-orders");
    return r.data;
  },
  async mappingDropdown(workCenterId?: string, utilityType?: UtilityType): Promise<MappingDropdown[]> {
    const r = await apiClient.get<MappingDropdown[]>("/api/v1/machine-utility/mappings/dropdown",
      { params: { work_center_id: workCenterId, utility_type: utilityType } });
    return r.data;
  },

  // ── Mapping master CRUD
  async listMappings(params?: Partial<MachineUtilityFilters>): Promise<MachineUtilityMapping[]> {
    const r = await apiClient.get<MachineUtilityMapping[]>("/api/v1/machine-utility/mappings", { params });
    return r.data;
  },
  async createMapping(data: MachineUtilityMappingCreate): Promise<MachineUtilityMapping> {
    const r = await apiClient.post<MachineUtilityMapping>("/api/v1/machine-utility/mappings", data);
    return r.data;
  },
  async updateMapping(id: string, data: MachineUtilityMappingUpdate): Promise<MachineUtilityMapping> {
    const r = await apiClient.patch<MachineUtilityMapping>(`/api/v1/machine-utility/mappings/${id}`, data);
    return r.data;
  },
  async deleteMapping(id: string): Promise<void> {
    await apiClient.delete(`/api/v1/machine-utility/mappings/${id}`);
  },

  // ── Consumption records CRUD
  async listRecords(params?: Partial<MachineUtilityFilters>): Promise<MachineConsumptionRecord[]> {
    const r = await apiClient.get<MachineConsumptionRecord[]>("/api/v1/machine-utility/records", { params });
    return r.data;
  },
  async createRecord(data: MachineConsumptionRecordCreate): Promise<MachineConsumptionRecord> {
    const r = await apiClient.post<MachineConsumptionRecord>("/api/v1/machine-utility/records", data);
    return r.data;
  },
  async updateRecord(id: string, data: MachineConsumptionRecordUpdate): Promise<MachineConsumptionRecord> {
    const r = await apiClient.patch<MachineConsumptionRecord>(`/api/v1/machine-utility/records/${id}`, data);
    return r.data;
  },
  async deleteRecord(id: string): Promise<void> {
    await apiClient.delete(`/api/v1/machine-utility/records/${id}`);
  },

  // ── Analytics
  async kpis(params?: Partial<MachineUtilityFilters>): Promise<MachineUtilityKPIs> {
    const r = await apiClient.get<MachineUtilityKPIs>("/api/v1/machine-utility/kpis", { params });
    return r.data;
  },
  async dailyTrend(params?: Partial<MachineUtilityFilters>): Promise<MachineUtilityTrendPoint[]> {
    const r = await apiClient.get<MachineUtilityTrendPoint[]>("/api/v1/machine-utility/trend/daily", { params });
    return r.data;
  },
  async breakdown(
    dimension: "machine" | "utility" | "shift" | "line" | "department",
    params?: Partial<MachineUtilityFilters>,
  ): Promise<MachineUtilityBreakdown> {
    const r = await apiClient.get<MachineUtilityBreakdown>("/api/v1/machine-utility/breakdown",
      { params: { ...params, dimension } });
    return r.data;
  },
};

File: backend/app/schemas/machine_utility.py
"""
Machine Utility Consumption Mapping Schemas
────────────────────────────────────────────
Two-layer design:

  MachineUtilityMapping  — master config: which meter serves which machine,
                            rated consumption, allocation share.
  MachineConsumptionRecord — operational log: actual consumption per machine
                              per date/shift, linked to production order/batch.

KPIs (computed in service):
  electricity_per_runtime_h    kWh / runtime_hour
  water_per_batch              m³  / batch
  air_per_batch                Nm³ / batch
  standby_consumption          sum of is_standby records per utility
  cost_per_batch               total_cost / distinct batch count
  cost_per_unit_produced       total_cost / produced quantity (where order linked)
  variance_pct                 (actual − expected) / expected × 100
"""
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, field_validator


# ══════════════════════════════════════════════════════════════════════════════
# MACHINE UTILITY MAPPING (master config)
# ══════════════════════════════════════════════════════════════════════════════

class MachineUtilityMappingCreate(BaseModel):
    machine_ref:     str
    machine_name:    str
    work_center_id:  Optional[UUID]    = None   # FK → work_centers.id
    department:      Optional[str]     = None
    production_line: Optional[str]     = None
    building_area:   Optional[str]     = None
    utility_type:    str               = "ELECTRICITY"
    device_id:       Optional[UUID]    = None   # FK → utility_devices.id (source meter)
    asset_id:        Optional[UUID]    = None   # FK → utility_assets.id
    allocation_pct:  Optional[Decimal] = None   # % share of shared meter
    rated_consumption: Optional[Decimal] = None # nameplate rate
    rated_unit:      Optional[str]     = None   # e.g. kW, m³/h, kg/h
    is_active:       bool              = True
    notes:           Optional[str]     = None

    @field_validator("utility_type")
    @classmethod
    def _chk_type(cls, v: str) -> str:
        valid = {
            "ELECTRICITY", "WATER", "SOFT_WATER", "PROCESS_WATER",
            "STEAM", "COMPRESSED_AIR", "SOLAR", "NATURAL_GAS",
            "WASTEWATER", "DIESEL", "CHILLED_WATER", "OTHER",
        }
        if v.upper() not in valid:
            raise ValueError(f"utility_type must be one of {valid}")
        return v.upper()


class MachineUtilityMappingUpdate(BaseModel):
    machine_name:    Optional[str]     = None
    work_center_id:  Optional[UUID]    = None
    department:      Optional[str]     = None
    production_line: Optional[str]     = None
    building_area:   Optional[str]     = None
    device_id:       Optional[UUID]    = None
    asset_id:        Optional[UUID]    = None
    allocation_pct:  Optional[Decimal] = None
    rated_consumption: Optional[Decimal] = None
    rated_unit:      Optional[str]     = None
    is_active:       Optional[bool]    = None
    notes:           Optional[str]     = None


class MachineUtilityMappingRead(BaseModel):
    id:              UUID
    machine_ref:     str
    machine_name:    str
    work_center_id:  Optional[UUID]    = None
    department:      Optional[str]     = None
    production_line: Optional[str]     = None
    building_area:   Optional[str]     = None
    utility_type:    str
    device_id:       Optional[UUID]    = None
    device_name:     Optional[str]     = None   # denormalized
    asset_id:        Optional[UUID]    = None
    allocation_pct:  Optional[Decimal] = None
    rated_consumption: Optional[Decimal] = None
    rated_unit:      Optional[str]     = None
    is_active:       bool
    notes:           Optional[str]     = None
    created_at:      Optional[datetime] = None
    updated_at:      Optional[datetime] = None

    model_config = {"from_attributes": True}


# ══════════════════════════════════════════════════════════════════════════════
# MACHINE CONSUMPTION RECORD (operational log)
# ══════════════════════════════════════════════════════════════════════════════

class MachineConsumptionRecordCreate(BaseModel):
    work_center_id:  UUID              # FK → work_centers.id (REQUIRED)
    record_date:     date
    utility_type:    str               = "ELECTRICITY"
    mapping_id:      Optional[UUID]    = None   # FK → machine_utility_mappings.id
    device_id:       Optional[UUID]    = None   # source meter FK
    shift_ref:       Optional[str]     = None
    department:      Optional[str]     = None
    production_line: Optional[str]     = None

    source_method:   str               = "MANUAL"
    is_estimated:    bool              = False
    is_standby:      bool              = False

    runtime_hours:   Optional[Decimal] = None

    nominal_rate:    Optional[Decimal] = None
    nominal_unit:    Optional[str]     = None

    actual_consumption: Optional[Decimal] = None
    actual_unit:     Optional[str]     = None

    production_order_id: Optional[UUID] = None
    batch_no:        Optional[str]     = None
    batch_lot_id:    Optional[UUID]    = None

    cost_rate:       Optional[Decimal] = None
    total_cost:      Optional[Decimal] = None
    currency_code:   Optional[str]     = "USD"

    is_anomaly:      bool              = False
    anomaly_note:    Optional[str]     = None
    notes:           Optional[str]     = None

    @field_validator("utility_type")
    @classmethod
    def _chk_type(cls, v: str) -> str:
        valid = {
            "ELECTRICITY", "WATER", "SOFT_WATER", "PROCESS_WATER",
            "STEAM", "COMPRESSED_AIR", "SOLAR", "NATURAL_GAS",
            "WASTEWATER", "DIESEL", "CHILLED_WATER", "OTHER",
        }
        if v.upper() not in valid:
            raise ValueError(f"utility_type must be one of {valid}")
        return v.upper()


class MachineConsumptionRecordUpdate(BaseModel):
    record_date:        Optional[date]     = None
    shift_ref:          Optional[str]      = None
    department:         Optional[str]      = None
    production_line:    Optional[str]      = None
    device_id:          Optional[UUID]     = None
    source_method:      Optional[str]      = None
    is_estimated:       Optional[bool]     = None
    is_standby:         Optional[bool]     = None
    runtime_hours:      Optional[Decimal]  = None
    nominal_rate:       Optional[Decimal]  = None
    nominal_unit:       Optional[str]      = None
    actual_consumption: Optional[Decimal]  = None
    actual_unit:        Optional[str]      = None
    variance_pct:       Optional[Decimal]  = None
    production_order_id: Optional[UUID]   = None
    batch_no:           Optional[str]      = None
    batch_lot_id:       Optional[UUID]     = None
    cost_rate:          Optional[Decimal]  = None
    total_cost:         Optional[Decimal]  = None
    currency_code:      Optional[str]      = None
    is_anomaly:         Optional[bool]     = None
    anomaly_note:       Optional[str]      = None
    notes:              Optional[str]      = None


class MachineConsumptionRecordRead(BaseModel):
    id:              UUID
    record_no:       str
    mapping_id:      Optional[UUID]    = None
    work_center_id:  UUID
    machine_ref:     str
    machine_name:    Optional[str]     = None
    utility_type:    str
    device_id:       Optional[UUID]    = None
    device_name:     Optional[str]     = None

    record_date:     date
    shift_ref:       Optional[str]     = None
    department:      Optional[str]     = None
    production_line: Optional[str]     = None

    source_method:   str
    is_estimated:    bool
    is_standby:      bool

    runtime_hours:   Optional[Decimal] = None

    nominal_rate:    Optional[Decimal] = None
    nominal_unit:    Optional[str]     = None

    actual_consumption: Optional[Decimal] = None
    actual_unit:     Optional[str]     = None
    variance_pct:    Optional[Decimal] = None

    production_order_id: Optional[UUID] = None
    production_order_no: Optional[str]  = None   # denormalized
    batch_no:        Optional[str]      = None
    batch_lot_id:    Optional[UUID]     = None

    cost_rate:       Optional[Decimal]  = None
    total_cost:      Optional[Decimal]  = None
    currency_code:   Optional[str]      = None

    is_anomaly:      bool
    anomaly_note:    Optional[str]      = None
    notes:           Optional[str]      = None

    entered_by_id:   Optional[UUID]     = None
    entered_by_name: Optional[str]      = None

    created_at:      Optional[datetime] = None
    updated_at:      Optional[datetime] = None

    model_config = {"from_attributes": True}


# ══════════════════════════════════════════════════════════════════════════════
# KPIs & ANALYTICS
# ══════════════════════════════════════════════════════════════════════════════

class MachineUtilityKPIs(BaseModel):
    """
    Aggregated machine utility KPIs over a date window.

    Reporting readiness flags:
      flag_high_variance   avg_variance_pct > 15 %
      flag_standby_waste   standby_pct > 20 % of total consumption
    """
    date_from: Optional[date] = None
    date_to:   Optional[date] = None

    machine_ref:  Optional[str] = None
    utility_type: Optional[str] = None

    # Volume totals
    total_consumption:    Decimal = Decimal("0")
    total_runtime_hours:  Optional[Decimal] = None
    total_standby_consumption: Optional[Decimal] = None
    consumption_unit:     Optional[str] = None

    # KPI 1 — electricity per runtime hour
    electricity_per_runtime_h: Optional[Decimal] = None   # kWh / h

    # KPI 2 — water per batch
    water_per_batch:      Optional[Decimal] = None   # m³ / batch

    # KPI 3 — compressed air per batch
    air_per_batch:        Optional[Decimal] = None   # Nm³ / batch

    # KPI 4 — standby consumption share
    standby_pct:          Optional[Decimal] = None   # %

    # KPI 5 — cost per batch
    cost_per_batch:       Optional[Decimal] = None

    # KPI 6 — cost per unit produced (if order linked)
    cost_per_unit:        Optional[Decimal] = None

    # General cost
    total_cost:           Optional[Decimal] = None
    avg_cost_rate:        Optional[Decimal] = None

    # Variance vs standard
    avg_variance_pct:     Optional[Decimal] = None   # deviation from nominal rate

    # Counts
    record_count:         int = 0
    batch_count:          int = 0
    order_count:          int = 0
    anomaly_count:        int = 0

    # Anomaly flags
    flag_high_variance:   bool = False   # avg_variance_pct > 15 %
    flag_standby_waste:   bool = False   # standby_pct > 20 %


class MachineUtilityTrendPoint(BaseModel):
    """Daily aggregation per machine per utility type."""
    date:              str     = ""
    machine_ref:       Optional[str] = None
    utility_type:      Optional[str] = None
    total_consumption: Decimal = Decimal("0")
    standby:           Decimal = Decimal("0")
    runtime_hours:     Decimal = Decimal("0")
    total_cost:        Optional[Decimal] = None
    batch_count:       int = 0
    record_count:      int = 0

    model_config = {"from_attributes": True}


class MachineUtilityBreakdownRow(BaseModel):
    """One row in a breakdown — by machine, utility type, shift, line, etc."""
    group_key:         Optional[str]
    label:             str
    total_consumption: Decimal = Decimal("0")
    total_cost:        Optional[Decimal] = None
    pct_of_total:      Optional[Decimal] = None
    runtime_hours:     Optional[Decimal] = None
    record_count:      int = 0


class MachineUtilityBreakdown(BaseModel):
    dimension:            str
    rows:                 List[MachineUtilityBreakdownRow]
    grand_total:          Decimal


# ── Dropdown helpers ───────────────────────────────────────────────────────────

class WorkCenterDropdown(BaseModel):
    id:             UUID
    work_center_id: str
    name:           str
    department:     Optional[str] = None
    type:           str

    model_config = {"from_attributes": True}


class MappingDropdown(BaseModel):
    id:           UUID
    machine_ref:  str
    machine_name: str
    utility_type: str
    rated_consumption: Optional[Decimal] = None
    rated_unit:   Optional[str] = None

    model_config = {"from_attributes": True}
